Modules/_pickle.c

```diff
@@ -6623,14 +6623,9 @@ _pickle_Unpickler_find_class_impl(UnpicklerObject *self,
         }
     }
 
-    module = PyImport_GetModule(module_name);
-    if (module == NULL) {
-        if (PyErr_Occurred())
-            return NULL;
-        module = PyImport_Import(module_name);
-        if (module == NULL)
-            return NULL;
-    }
+    module = PyImport_Import(module_name);
+    if (module == NULL)
```

Comment:
It seems to me that the difference in behavior here occurs when `PyImport_Import` returns `NULL` *without* raising an exception. I'm not really sure why this would occur. Is it safe to simply propagate the `NULL` up to the next level here?

Reply:
IIRC (a big if), the issue wasn't that `PyImport_Import` returned NULL, it's that `PyImport_GetModule` can return a module object where the body of the module is still being executed (in another thread), and this partially initialised thing is then used immediately by the thread doing the unpickling.

Reply:
I'm not sure how this fixes things, though. It seems like the only change in behavior is for the case where `PyImport_Import` returns `NULL` without raising an exception. Why would that ever happen, and why was it being explicitly handled here in the first place?

Reply:
In the failing case before, `PyImport_Import` was never called: `GetModule` returned non-NULL, and this object fell through to the `getattribute`, which then (sometimes) failed because that attribute did not exist yet.

On L6626 I changed `_GetModule` -> `_Import`.   This, I believe, invokes the locking logic described here https://github.com/python/cpython/blob/cd90f6a3692e0f7ef0a13aae651e19a08d1f9b31/Python/import.c#L145 which prevents the original issue.

Reply:
My mistake, I didn't realize that you were changing the import function. I'm still concerned as to why this was added in the first place, and whether this is a wider problem.

Reply:
Another nitpick, per [PEP 7](https://www.python.org/dev/peps/pep-0007/):

> Code structure: one space between keywords like if, for and the following left paren; no spaces inside the paren; braces are required everywhere, even where C permits them to be omitted, but do not add them to code you are not otherwise modifying. All new C code requires braces.

So can you add braces around this?